### PR TITLE
Add checking if the source repository is clean

### PR DESCRIPTION
**Pull request by slimreaper35**
STONEBLD-465

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] New code has type annotations
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

File: cachi2/core/errors.py [hunk 1]
```diff
@@ -125,6 +125,20 @@ def friendly_msg(self) -> str:
         return _friendly_error_msg(str(self), self.notice)
 
 
+class UncleanSourceRepo(Cachi2Error):
+    """If that repo is unclean, Cachi2 will not process the repo and will fail the run."""
+
+    is_invalid_usage: ClassVar[bool] = True
+    solution = (
+        "Fetching dependencies is not supported for unclean git repositories.\n"
+        "Use `git status` to check if the source repository is clean."
```
[the remark this answers is not in the document]
Comment (slimreaper35) [reply cachi2/core/errors.py]: Should we log only untracked files or all files that are not commited ?

[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/errors.py]: IMO we should log those that caused the the repo to be considered dirty. If I understand it correctly, that's all untracked and uncommitted (staged or not staged) files.

Something like `git status --porcelain` could be nice

File: cachi2/interface/cli.py [hunk 2]
```diff
@@ -321,3 +325,17 @@ def _get_build_config(output_dir: Path) -> BuildConfig:
...
+
+def _is_dirty_git_repository(source_dir: Path) -> bool:
+    try:
+        repo = Repo(source_dir)
+    except InvalidGitRepositoryError:
+        # not a git repository => clean
+        return False
+
+    if repo.is_dirty(untracked_files=True):
+        log.info(repo.git.status("--porcelain"))
```
Comment (chmeliik) [inline cachi2/interface/cli.py:338]: Currently the output looks like this:

```
❯ cachi2 fetch-deps pip
2023-03-30 11:39:24,910 INFO  M tox.ini
?? .markdownlint.yaml
?? tests/unit/data/gomod-mocks/
2023-03-30 11:39:24,910 ERROR UncleanSourceRepo: The repository is considered dirty
Error: UncleanSourceRepo: The repository is considered dirty
  Fetching dependencies is not supported for unclean git repositories.
  The source repository must be clean (no untracked files, no changes).
```

Let's make sure the git status output will be properly aligned and log something to explain what the output is. And we should probably log this at ERROR level, @fepas can confirm :slightly_smiling_face: 

```
2023-03-30 11:39:24,910 ERROR The repository is not clean (git status --porcelain):
 M tox.ini
?? .markdownlint.yaml
?? tests/unit/data/gomod-mocks/
```

Comment (slimreaper35) [reply cachi2/interface/cli.py:338]: What do you think about this output ?
(Skip temporary ugly functional programming)

```
log.error(f"Changes to be committed: {list(map(lambda x: x.a_path, list(repo.index.diff('HEAD'))))}")
log.error(f"Changes not staged for commit: {list(map(lambda x: x.a_path, list(repo.index.diff(None))))}")
log.error(f"Untracked files: {repo.untracked_files}")
```

Comment (chmeliik) [reply cachi2/interface/cli.py:338]: IMO one file per line is easier to read, git status is pretty good in this regard. For my part, as long as the git status output is separated by a newline from the log message, it's readable enough

**Comment by lkolacek** [comment]
> @team what do you guys think, how long until the rhtap build pipeline introduces a corner case where it _will_ modify the repo before the fetch-deps task?
> 
> Should we make this check configurable right away? On the user side or the "server" (confg file) side? Should we even have the check, or we should we report dirtiness in the SBOM somehow?

That's a great point, Adam.

Well, IMO failing the build here if the repo is unclean is the best option still. +1 to add config option on "server" side to disable it (if some corner case shows up and it will be approved).

Because if we'll make this check Warning log only .. nobody will notice that something might be wrong, and I would not probably polute SBOM with such info if nobody requested it.

**Comment by brunoapimentel** [comment]
> > @team what do you guys think, how long until the rhtap build pipeline introduces a corner case where it _will_ modify the repo before the fetch-deps task?
> > Should we make this check configurable right away? On the user side or the "server" (confg file) side? Should we even have the check, or we should we report dirtiness in the SBOM somehow?
> 
> That's a great point, Adam.
> 
> Well, IMO failing the build here if the repo is unclean is the best option still. +1 to add config option on "server" side to disable it (if some corner case shows up and it will be approved).
> 
> Because if we'll make this check Warning log only .. nobody will notice that something might be wrong, and I would not probably polute SBOM with such info if nobody requested it.

In the end, Cachi2 would always download and report the dependencies as they are currently in the source code, even it that doesn't match with the last commit/version.  I tend to think that adding a qualifier to the package version (such as `1.0.1+unclean`) and a WARNING log should be enough.

In the current state of RHTAP, this scenario is not possible (unless the user modifies the pipeline).

**Comment by chmeliik** [comment]
> In the end, Cachi2 would always download and report the dependencies as they are currently in the source code, even it that doesn't match with the last commit/version. I tend to think that adding a qualifier to the package version (such as `1.0.1+unclean`) and a WARNING log should be enough.
> 
> In the current state of RHTAP, this scenario is not possible (unless the user modifies the pipeline).

This doesn't even affect most dependencies, only the top-level packages (from the user's repo) and dependencies from file paths.

And the `gomod-vendor-check` flag, which a) needs a rework and b) arguably still does what it says (although the modifications are compared with the last clean commit, not with the state before calling `fetch-deps`)

**Comment by chmeliik** [comment]
> This doesn't even affect most dependencies, only the top-level packages (from the user's repo) and dependencies from file paths.

As in, the purl we (will) use to identify the top-level package and local dependencies will not be a 100% accurate representation of the source code

**Comment by chmeliik** [comment]
And you know what, checking this at the start of the fetch-dependencies task is not sufficient anyway. In fact, the fetch-dependencies task itself can modify the repo - sometimes intentionally (`cachi2 inject-files`), sometimes not (any `go` command).

We can add this check and be happy that the SBOM was still accurate at the time we generated it, but we have no idea what happens between fetch-dependencies and the actual build.

I think the only way to look at this is that modifications made by the build pipeline are an internal implementation detail - clean repo goes in, container image comes out, what happens inside doesn't matter. And if a different part of the pipeline makes a change that *should* matter, it's up to that other part to report that.

**Comment by chmeliik** [comment]
TL;DR let's drop this and not worry about repo modifications on the cachi2 level?

(we might still want to worry about some of those that cachi2 does unintentionally, but that's a different story)

**Comment by brunoapimentel** [comment]
> We can add this check and be happy that the SBOM was still accurate at the time we generated it, but we have no idea what happens between fetch-dependencies and the actual build.

Good point here. Agree, let's drop this.

**Comment by chmeliik** [comment]
@slimreaper35 we won't be needing this after all, apologies for throwing away the effort spent :sweat_smile:
